transports/httpclient/src/main/java/org/commonjava/maven/galley/transport/htcli/UploadMetadataGenTransferDecorator.java

```diff
@@ -0,0 +1,138 @@
+import org.commonjava.maven.galley.event.EventMetadata;
+import org.commonjava.maven.galley.io.AbstractTransferDecorator;
+import org.commonjava.maven.galley.model.SpecialPathInfo;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.model.TransferOperation;
+import org.commonjava.maven.galley.spi.io.SpecialPathManager;
+import org.commonjava.maven.galley.transport.htcli.model.HttpExchangeMetadata;
+import org.commonjava.maven.galley.transport.htcli.model.HttpExchangeMetadataFromRequestHeader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.inject.Alternative;
+import javax.inject.Named;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+
+import static org.commonjava.maven.galley.spi.cache.CacheProvider.STORE_HTTP_HEADERS;
+
+@Named
+@Alternative
+public class UploadMetadataGenTransferDecorator
+        extends AbstractTransferDecorator
+{
+    private static final Logger logger = LoggerFactory.getLogger( UploadMetadataGenTransferDecorator.class );
+
+    private SpecialPathManager specialPathManager;
+
+    public UploadMetadataGenTransferDecorator( SpecialPathManager specialPathManager )
+    {
+        this.specialPathManager = specialPathManager;
+    }
+
+    @Override
+    public OutputStream decorateWrite( OutputStream stream, Transfer transfer, TransferOperation op,
+                                       EventMetadata metadata )
+            throws IOException
+    {
+        final SpecialPathInfo specialPathInfo = specialPathManager.getSpecialPathInfo( transfer, metadata.getPackageType() );
+
+        final Boolean isMetadata = specialPathInfo != null && specialPathInfo.isMetadata();
+        final Boolean isUpload = op == TransferOperation.UPLOAD;
+        final Boolean hasRequestHeaders =
+                metadata.get( STORE_HTTP_HEADERS ) != null && metadata.get( STORE_HTTP_HEADERS ) instanceof Map;
+
+        // We only care about non-metadata artifact uploading for http-metadata generation
+        if ( isUpload && !isMetadata && hasRequestHeaders )
+        {
+            @SuppressWarnings( "unchecked" ) Map<String, List<String>> storeHttpHeaders =
+                    (Map<String, List<String>>) metadata.get( STORE_HTTP_HEADERS );
+            writeMetadata( transfer, new ObjectMapper(), storeHttpHeaders );
+        }
+
+        return super.decorateWrite( stream, transfer, op, metadata );
+    }
+
+    private void writeMetadata( final Transfer target, final ObjectMapper mapper, final Map<String, List<String>> requestHeaders )
+    {
+        Transfer metaTxfr = target.getSiblingMeta( HttpExchangeMetadata.FILE_EXTENSION );
+        if ( metaTxfr == null )
+        {
+            if ( target.isDirectory() )
+            {
+                logger.debug( "DIRECTORY. Using HTTP exchange metadata file INSIDE directory called: {}", HttpExchangeMetadata.FILE_EXTENSION );
```

Comment:
Does this only happen if someone is trying to write to a directory as if it was a file? I'm tempted to say we should log that as an error, since it's insane...but that's something the I guess the CacheProvider should throw the error for?

I'd revise this one to trace()